<a href="https://colab.research.google.com/github/rahiakela/data-science-bookcamp-ten-case-studies/blob/CASE-STUDY-1-FINDING-THE-WINNING-STRATEGY-IN-A-CARD-GAME/1_computing_probabilities_using_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Computing Probabilities Using Python

Few things in life are certain; most things are driven by chance.guaranteed. Randomness permeates our day to day experiences. Fortunately, that randomness can still be mitigated and controlled. 

We know that some unpredictable events occur more rarely than others, and that certain decisions carry less uncertainty than other much more risky choices.

We can intrinsically sense these trade-offs in certainty because even the most unpredictable systems still show some predictable behaviors. These behaviors have been rigorously studied using probability theory. Probability
theory is an inherently complex branch of math. However, aspects of the theory can be understood without knowing the mathematical underpinnings.

In fact, difficult probability problems can be solved in Python without needing to know a single math equation. Such an equation-free approach to probability requires a baseline understanding of what mathematicians call a sample space.



## Sample Space Analysis: An Equation-Free Approach for Measuring Uncertainty in Outcomes

Certain actions have measurable outcomes. A sample space is the set of all the possible outcomes
that an action could produce. Lets take the simple action of flipping a coin. The coin will land on
either heads or tails. Thus, the coin-flip will produce one of 2 measurable outcomes: 'Heads' or
'Tails'.

In [0]:
sample_space = {'Heads', 'Tails'}

Each element occupies an equal fraction of space within the set. Therefore, we expect 'Heads' to be selected with a frequency of 1/2. That frequency is formally defined as the probability of an outcome.
All outcomes within sample_space share an identical probability, which is equal to 1 / len(sample_space).

In [2]:
probability_heads = 1 / len(sample_space)
print(f'Probability of choosing heads is {probability_heads}')

Probability of choosing heads is 0.5


The probability of choosing 'Heads' equals 0.5. This relates directly to the action of flipping a
coin. We’ll assume the coin is unbiased, which means the coin is equally likely to fall on either
heads or tails. Thus, a coin-flip is conceptually equivalent to choosing a random element from
sample_space. The probability of the coin landing on heads is therefore 0.5. The probability of
it landing on tails is also equal to 0.5.

An event is the subset of those elements within sample_space that satisfy some event condition
. An event condition is a simple Boolean function whose input is a single sample_space
element. The function returns True only if the element satisfies our condition constraints.

<img src='https://github.com/rahiakela/img-repo/blob/master/coin-flipping-1.JPG?raw=1' width='800'/>

Lets define 2 event conditions: one where the coin lands on either heads or tails, and another
where the coin lands on neither heads nor tails.

In [0]:
def is_heads_or_tails(outcome):
  return outcome in {'Heads', 'Tails'}

def is_neither(outcome):
  return not is_heads_or_tails(outcome)  

Also, for completeness-sake, lets define event conditions for the 2 basic events in which the coin satisfies exactly one of our 2 potential outcomes.

In [0]:
def is_heads(outcome):
  return outcome == 'Heads'

def is_tails(outcome):
  return outcome == 'Tails'  

We can pass event conditions into a generalized function. That get_events function is defined below. Its inputs are an event condition, and also a generic sample space. The function iterates through generic_sample_space and returns the set of outcomes where event_condition(outcome) is True.

In [0]:
def get_event(event_condition, sample_space):
  return set([outcome for outcome in sample_space if event_condition(outcome)])

Lets execute get_event on our 4 event conditions. Afterwards, we’ll output the 4 extracted events.

In [6]:
event_conditions = [is_heads_or_tails, is_heads, is_tails, is_neither]

for event_condition in event_conditions:
  print(f'Event Condition: {event_condition.__name__}')
  event = get_event(event_condition, sample_space)
  print(f'Event: {event}\n')

Event Condition: is_heads_or_tails
Event: {'Heads', 'Tails'}

Event Condition: is_heads
Event: {'Heads'}

Event Condition: is_tails
Event: {'Tails'}

Event Condition: is_neither
Event: set()



This property can be generalized to include multi-element
events. The probability of an event is equal to len(event) / len(sample_space), but only if
all outcomes are known to occur with equal likelihood. In other words, the probability of a
multi-element event for a fair coin is equal to event size divided by the sample space size.

We’ll now leverage event size to compute the 4 event probabilities.

In [7]:
def compute_probability(event_condition, generic_sample_space):
  event = get_event(event_condition, generic_sample_space)  # extract the event associated with an inputted event condition in order to compute its probability
  return len(event) / len (generic_sample_space)            # Probability is equal to event size divided by sample space size.

for event_condition in event_conditions:
  prob = compute_probability(event_condition, sample_space)
  name = event_condition.__name__
  print(f'Probability of event arising from "{name}" is {prob}')  

Probability of event arising from "is_heads_or_tails" is 1.0
Probability of event arising from "is_heads" is 0.5
Probability of event arising from "is_tails" is 0.5
Probability of event arising from "is_neither" is 0.0


The executed code outputs a diverse range of event probabilities, the lowest of which is 0.0 and the largest of which is 1.0. 

These values represent the lower and upper bounds of probability; no
probability can ever fall below 0.0 or rise above 1.0.

### Analyzing a Biased Coin

We computed probabilities for an unbiased coin. What would happen if that coin was biased?
Suppose, for instance, that a coin is 4 times more likely to land on heads relative to tails. How do
we compute the likelihoods of outcomes that are not weighted in an equal manner? Simple!
We’ll construct a weighed sample space, represented by a Python dictionary. Each outcome will
be treated as a key whose value maps to the associated weight. 

In our example, heads is weighted
4 times as heavily as tails. Therefore, we’ll map 'Tails' to 1 and 'Heads' to 4.

In [0]:
weighted_sample_space = {'Heads': 4, 'Tails': 1}

Our new sample space is stored within a dictionary. This allows us to redefine sample-space size
as the sum of all dictionary weights. Within weighted_sample_space, that sum will equal 5.

In [0]:
sample_space_size = sum(weighted_sample_space.values())
assert sample_space_size == 5

We can redefine event size in similar manner. Each event is a set of outcomes. These outcomes
map to weights. Summing over these weights will yield the event size. 

Thus, the size of the event
satisfying the is_heads_or_tails event condition is also 5.

In [0]:
event = get_event(is_heads_or_tails, weighted_sample_space)
event_size = sum(weighted_sample_space[outcome] for outcome in event)
assert event_size == 5

Our generalized definitions of sample-space size and event size permit us to create a
compute_event_probability function. The function takes as input a generic_sample_space
variable that can be either a weighted dictionary or an unweighted set.

In [0]:
def compute_weighted_probability(event_condition, generic_sample_space):
  event = get_event(event_condition, generic_sample_space)
  if type(generic_sample_space) == type(set()):    # Checks if generic_event_space is a set.
    return len(event) / len(generic_sample_space)

  event_size = sum(generic_sample_space[outcome] for outcome in event)
  return event_size / sum(generic_sample_space.values())  

We can now output all the event probabilities for the biased coin without needing to redefine our
4 event condition functions.

In [12]:
for event_condition in event_conditions:
  prob = compute_weighted_probability(event_condition, weighted_sample_space)
  name = event_condition.__name__
  print(f'Probability of event arising from "{name}" is {prob}') 

Probability of event arising from "is_heads_or_tails" is 1.0
Probability of event arising from "is_heads" is 0.8
Probability of event arising from "is_tails" is 0.2
Probability of event arising from "is_neither" is 0.0


With just few lines of code, we have constructed a tool for solving many problems in probability.

## Computing Non-Trivial Probabilities

### Problem 1: Analyzing a Family with 4 Children

Suppose a family has 4 children. What is the probability that exactly 2 of the children are boys?
We’ll assume that each child is equally likely to be either a boy or a girl. This allows us to
construct an unweighted sample space where each outcome is a 4-element tuple representing one
possible sequence of 4 children.

<img src='https://github.com/rahiakela/img-repo/blob/master/4-successive-children.JPG?raw=1' width='800'/>

The sample space for 4 successive children. Each row in the sample space
contains one of 16 possible outcomes. Every outcome represents a unique combination
of 4 children. The sex of each child is signaled by a letter; B for Boy and G for Girl.
Outcomes with 2 boys are marked by an arrow. There are 6 such arrows present. Thus,
the probability of 2 boys equals 6 / 16.

In [13]:
possible_children = ['Boy', 'Girl']
sample_space = set()

for child_1 in possible_children:
  for child_2 in possible_children:
    for child_3 in possible_children:
      for child_4 in possible_children:
        outcome = (child_1, child_2, child_3, child_4)
        sample_space.add(outcome)

print(f'sample_space : {sample_space}')

sample_space : {('Boy', 'Boy', 'Girl', 'Girl'), ('Boy', 'Girl', 'Boy', 'Girl'), ('Girl', 'Girl', 'Boy', 'Boy'), ('Girl', 'Girl', 'Girl', 'Girl'), ('Boy', 'Girl', 'Girl', 'Girl'), ('Boy', 'Boy', 'Boy', 'Boy'), ('Boy', 'Girl', 'Girl', 'Boy'), ('Girl', 'Girl', 'Girl', 'Boy'), ('Boy', 'Boy', 'Girl', 'Boy'), ('Boy', 'Girl', 'Boy', 'Boy'), ('Girl', 'Boy', 'Girl', 'Girl'), ('Girl', 'Boy', 'Girl', 'Boy'), ('Girl', 'Girl', 'Boy', 'Girl'), ('Girl', 'Boy', 'Boy', 'Boy'), ('Girl', 'Boy', 'Boy', 'Girl'), ('Boy', 'Boy', 'Boy', 'Girl')}


We ran 4 nested for-loops to explore the sequence of 4 births. This is not an efficient use of code.
We can more easily generate our sample space using Python’s build-in itertools.product
function. The function returns all pairwise combinations of all elements across all input lists.

In [0]:
from itertools import product

'''
The * operator unpacks multiple arguments stored within a list. These arguments
are then passed into a specified function. Thus, calling ` product(*(4 *
[possible_children]))` is equivalent to calling product(possible_children,
possible_children, possible_children, possible_children).
'''
all_combinations = product(*(4 * [possible_children]))

assert set(all_combinations) == sample_space

We can make our code even more efficient by executing set(product(possible_children,
repeat=4)). In general, running product(possible_children, repeat=n) will return an
iterable over all possible combinations of n children.

In [0]:
sample_space_efficient = set(product(possible_children, repeat=4))

assert sample_space == sample_space_efficient

Lets calculate the fraction of sample_space that is composed of families with 2 boys. We’ll first
define a has_two_boys event condition. Afterwards, will pass that condition into
compute_event_probability.

In [16]:
def has_two_boys(outcome):
  return len([child for child in outcome if child == 'Boy']) == 2

prob = compute_probability(has_two_boys, sample_space)
print(f'Probability of 2 boys is {prob}')  

Probability of 2 boys is 0.375


The probability of exactly 2 boys being born in family of 4 children is 0.375. By implication, we
expect 37.5% of families with 4 children to contain an equal number of boys and girls. Of
course, the actual observed percentage of families with 2 boys will vary due to random chance.

### Problem 2: Analyzing Multiple Dice Rolls

Suppose we’re shown a fair six-sided die whose faces are numbered from 1 to 6. The die is
rolled 6 times. What is the probability that these 6 dice-rolls add up to 21?

We’ll begin by defining the possible values of any single roll. These are integers that range from
1 to 6.

In [27]:
possible_rolls = list(range(1, 7))
print(f'possible_rolls : {possible_rolls}')

possible_rolls : [1, 2, 3, 4, 5, 6]


Next, we’ll create the sample space for 6 consecutive rolls using the product function.

In [30]:
sample_space = set(product(possible_rolls, repeat=6))
print(sample_space)

{(6, 1, 3, 3, 2, 6), (1, 3, 4, 4, 2, 3), (1, 4, 6, 1, 3, 3), (3, 2, 1, 3, 6, 5), (3, 6, 6, 1, 1, 5), (2, 6, 3, 5, 6, 4), (5, 3, 3, 6, 1, 6), (5, 6, 2, 6, 6, 1), (2, 3, 3, 2, 4, 1), (3, 6, 3, 1, 6, 5), (5, 6, 5, 5, 5, 3), (1, 3, 6, 5, 4, 5), (3, 3, 5, 3, 5, 5), (2, 5, 5, 1, 2, 5), (4, 1, 5, 5, 6, 5), (5, 3, 3, 4, 3, 2), (5, 6, 2, 4, 4, 5), (3, 6, 3, 3, 4, 1), (6, 2, 4, 4, 4, 5), (6, 1, 1, 3, 2, 1), (6, 5, 6, 1, 5, 1), (5, 4, 5, 3, 4, 5), (3, 3, 3, 2, 4, 2), (5, 6, 4, 1, 1, 3), (2, 5, 3, 2, 4, 3), (3, 5, 5, 5, 3, 6), (3, 1, 1, 3, 5, 3), (4, 2, 6, 4, 4, 3), (5, 4, 5, 1, 6, 1), (2, 5, 2, 4, 6, 6), (6, 2, 6, 5, 6, 3), (2, 3, 4, 5, 6, 4), (2, 3, 6, 6, 2, 4), (2, 5, 1, 4, 2, 4), (1, 4, 1, 3, 6, 2), (4, 1, 6, 6, 4, 6), (1, 3, 3, 5, 1, 4), (2, 3, 1, 5, 1, 4), (2, 6, 5, 2, 2, 4), (1, 3, 2, 5, 1, 6), (3, 1, 3, 5, 3, 2), (1, 1, 1, 2, 6, 3), (2, 5, 6, 6, 1, 4), (4, 1, 1, 5, 3, 6), (6, 3, 4, 5, 2, 3), (6, 6, 6, 4, 5, 4), (2, 6, 6, 6, 6, 2), (5, 4, 2, 4, 2, 2), (2, 2, 3, 3, 2, 5), (3, 5, 2, 3, 3, 1),

Finally, we’ll define a has_sum_of_21 event condition that we’ll subsequently pass into
compute_event_probability.

In [32]:
def has_sum_of_21(outcome):
  return sum(outcome) == 21

# prob = compute_weighted_probability(lambda x: x == 21, weighted_sample_space)
prob = compute_weighted_probability(has_sum_of_21, sample_space)
print(prob)
assert prob == compute_probability(has_sum_of_21, sample_space)
print(f'Probability of dice summing to 21 is {prob}')

0.09284979423868313
Probability of dice summing to 21 is 0.09284979423868313


### Problem 3: Computing Dice-Roll Probabilities using Weighted Sample Spaces

We’ve just computed the likelihood of 6 rolled dice summing to 21. Now, lets recompute that
probability using a weighted sample space. How do we convert our unweighted sample-space set
into a weighted sample-space dictionary? 

The solution is simple. We must first identify all
possible sums of 6 dice-rolls. Then, we must count the number of times each sum appears across
all possible 6-dice-roll-combinations. These combinations are already stored in our computed
sample_space set. 

By mapping the dice-roll sums to their occurrence counts, we will produce a
weighted_sample_space result.

In [0]:
from collections import defaultdict

weighted_sample_space = defaultdict(int)

for outcome in sample_space:            # Each outcome contains a unique combination of six rolled dice.
  total = sum(outcome)                  # We compute the summed value of 6 unique dice rolls.
  weighted_sample_space[total] += 1     # We update the occurrence count for a summed dice value.

Before we recompute our probability, lets briefly explore the properties of
weighted_sample_space. Not all weights in the sample space are equal. Some of the weights are much smaller than others. 

For instance, there is only one way for the dice to sum to 6. We
must roll precisely 6 ones to achieve that dice-sum combination. Hence, we expect weighted_sample_space[6] to equal 1. Furthermore, we expect
weighted_sample_space[36] to also equal 1, since we must roll 6 sixes to achieve a sum of 36.

In [0]:
assert weighted_sample_space[6] == 1
assert weighted_sample_space[36] == 1

Meanwhile, the value of weighted_sample_space[21] is noticeably higher.

In [35]:
num_combinations = weighted_sample_space[21]
print(f'There are {num_combinations} ways for six rolled dice to sum to 21')

There are 4332 ways for six rolled dice to sum to 21


There are 4332 ways for 6 rolled dice to sum to 21. 

For example, we could roll 4 fours, followed
by a three and then a two. Or we could roll 3 fours followed by a five, a three, and a one.
Thousands of other combinations are possible. This is why a sum of 21 is much more probable
than a sum of 6.

In [0]:
assert sum([4, 4, 4, 4, 3, 2]) == 21
assert sum([4, 4, 4, 5, 3, 1]) == 21

Please note that the observed count of 4332 is equal to the length of an unweighted event whose
dice-rolls add up to 21. Also, the sum of values in weighted_sample is equal to the length of
sample_space.

In [0]:
event = get_event(has_sum_of_21, sample_space)

assert weighted_sample_space[21] == len(event)
assert sum(weighted_sample_space.values()) == len(sample_space)

Lets now recompute the probability using the dictionary. weighted_sample_space The final
probability of rolling a 21 should remain unchanged.

In [43]:
prob = compute_weighted_probability(lambda x: x == 21, weighted_sample_space)

assert prob == compute_weighted_probability(has_sum_of_21, sample_space)
print(f'Probability of dice summing to 21 is {prob}')

Probability of dice summing to 21 is 0.09284979423868313


What is the benefit of using a weighted sample space over an unweighted one? Less memory
usage! As we see below, the unweighted sample_space set has on the order of 1500x more
elements than the weighted sample space dictionary.

In [44]:
print('Number of Elements in Unweighted Sample Space:')
print(len(sample_space))
print('Number of Elements in Weighted Sample Space:')
print(len(weighted_sample_space))

Number of Elements in Unweighted Sample Space:
46656
Number of Elements in Weighted Sample Space:
31


## Computing Probabilities Over Interval Ranges